engine/consensus/approvals/assignment_collector.go

```diff
@@ -0,0 +1,354 @@
+// - verification node identity
+// - attestation signature
+// - signature of verification node
+// - chunk index sanity check
+// - block ID sanity check
+// Returns:
+// - engine.InvalidInputError - result approval is invalid
+// - exception in case of any other error, usually this is not expected
+// - nil on successful check
+func (ac *AssignmentCollector) validateApproval(approval *flow.ResultApproval) error {
+	// check that approval is for the expected result to reject incompatible inputs
+	if approval.Body.ExecutionResultID != ac.ResultID {
+		return fmt.Errorf("this AssignmentCollector processes only approvals for result (%x) but got an approval for (%x)", ac.ResultID, approval.Body.ExecutionResultID)
+	}
+
+	// approval has to refer same block as execution result
+	if approval.Body.BlockID != ac.BlockID() {
+		return engine.NewInvalidInputErrorf("result approval for invalid block, expected (%x) vs (%x)",
+			ac.BlockID(), approval.Body.BlockID)
+	}
+
+	chunkIndex := approval.Body.ChunkIndex
+	if chunkIndex >= uint64(ac.result.Chunks.Len()) {
+		return engine.NewInvalidInputErrorf("chunk index out of range: %v", chunkIndex)
+	}
+
+	identity, found := ac.authorizedApprovers[approval.Body.ApproverID]
+	if !found {
+		return engine.NewInvalidInputErrorf("approval not from authorized verifier")
+	}
+
+	err := ac.verifyAttestationSignature(&approval.Body, identity)
+	if err != nil {
+		return fmt.Errorf("validating attestation signature failed: %w", err)
+	}
+
+	err = ac.verifySignature(approval, identity)
+	if err != nil {
+		return fmt.Errorf("validating approval signature failed: %w", err)
+	}
+
+	return nil
+}
+
+func (ac *AssignmentCollector) ProcessApproval(approval *flow.ResultApproval) error {
+	err := ac.validateApproval(approval)
+	if err != nil {
+		return fmt.Errorf("could not validate approval: %w", err)
+	}
+
+	if cached := ac.verifiedApprovalsCache.Get(approval.Body.PartialID()); cached != nil {
+		// we have this approval cached already, no need to process it again
+		return nil
+	}
+
+	ac.verifiedApprovalsCache.Put(approval)
+
+	for _, collector := range ac.allCollectors() {
+		err := collector.ProcessApproval(approval)
+		if err != nil {
+			return fmt.Errorf("could not process approval: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (ac *AssignmentCollector) RequestMissingApprovals(maxHeightForRequesting uint64) error {
+	for _, collector := range ac.allCollectors() {
+		if collector.IncorporatedBlock().Height > maxHeightForRequesting {
+			continue
+		}
+
+		for chunkIndex, verifiers := range collector.CollectMissingVerifiers() {
+			// Retrieve information about requests made for this chunk. Skip
+			// requesting if the blackout period hasn't expired. Otherwise,
+			// update request count and reset blackout period.
+			requestTrackerItem := ac.requestTracker.Get(ac.ResultID, collector.IncorporatedBlockID(), chunkIndex)
+			if requestTrackerItem.IsBlackout() {
+				continue
+			}
+			requestTrackerItem.Update()
```

Comment:
would that be useful to merge `IsBlockout` and `Update` into one atomic function `TryUpdate` that returns a bool indicating whether it was black out?

Not necessary in this PR

Reply:
I started adding this but saw that we still need request tracker for logging